Manual on the standard TLA+ operators (#339)
informalsystems/apalache: docs/lang/standard-operators.md

[Community Modules].

We explain the semantics of the operators under the lenses of the
[Apalache model checker].  Traditionally, the emphasis was put on the temporal
operators and action operators, as they build the foundation of TLA. We focus
on the "+" aspect of the language, as this part of the language is absolutely
necessary for writing and reading system specifications.  Moreover, we treat
equally the "core" operators of TLA+ and the "library" operators: This
distinction is less important to the language users than to the tool developers.

In this document, we present the semantics of TLA+, as if it was executed on a
computer that is equipped with an additional device that we call an _oracle_.
Most of the TLA+ operators are understood as deterministic operators, so they
can be executed on your computer. A few operators are non-deterministic, so
they require the oracle to resolve non-determinism, see [Control Flow and
Non-determinism]. This is one of the most important features that makes TLA+
distinct from programming languages.  Wherever possible, we complement the
English semantics with code in [Python](https://www.python.org/). Although our

> semantics is more restrictive than the denotational semantics in Chapter 16 of
> [Specifying Systems], it is very close to the treatment of TLA+ by the model

Reviewer: ```suggestion
semantics are more restrictive than the denotational semantics in Chapter 16 of
[Specifying Systems], they are very close to the treatment of TLA+ by the model
```

      [`i^k`](./integers.md#pow)
 - Integer predicates:
    - [`i < k`](./integers.md#lt), [`i > k`](./integers.md#gt),
    - [`i <= k`](./integers.md#lte) (also `i =< k` and `i \leq k`), 
    - [`i => k`](./integers.md#gte) (also `i >= k` and `i \geq k`)
 - Integer sets: [`i..k`](./integers.md#range),
    [`Int`](./integers.md#const), [`Nat`](./integers.md#const)

### Strings :abcd:

_String constants_. You learned it!

 - String literals, e.g., `"hello"` and `"TLA+ is awesome"`
 - Set of all finite strings: `STRING`

### Sets :sushi:

_Like frozen sets in Python, but cooler_ [Learn more...](./sets.md)

 - Set algebra:
   - [`S \union T`](./sets.md#union) (also `S \cup T`),
   - [`S \intersect T`](./sets.md#intersect) (also `S \cap T`),
   - [`S \ T`](./sets.md#setminus)
 - Set predicates:
    - [`x \in S`](./sets.md#in) and [`x \notin S`](./sets.md#notin),
    - [`S \subset T`](./sets.md#subset),
        [`S \subseteq T`](./sets.md#subseteq),
        [`S \supset T`](./sets.md#supset),
        [`S \supseteq T`](./sets.md#supseteq)
 - Set filter: [`{ x \in S: p }`](./sets.md#filter)
 - Set map: [`{ e: x \in S }`](./sets.md#map)

>  - Powers: [`SUBSET S`](./sets.md#powerset) and [`UNION S`](./sets.md#fold)

Author: That makes sense. I also realized that I have missed the basic set constructor `{e_1, ..., e_n}` in this list ;-)

 - Bounded quantifiers:
    [`\A x \in S: p`](./logic.md#forallBounded) and [`\E x \in S: p`](./logic.md#existsBounded)
 - Unbounded quantifiers:
    [`\A x: p`](./logic.md#forall) and [`\E x: p`](./logic.md#exists)
 - Choice:
    [`CHOOSE x \in S: p`](./logic.md#chooseBounded) and [`CHOOSE x: p`](./logic.md#choose)

### Functions :chart:

_Like frozen dictionaries in Python, but cooler_. [Learn more...](./functions.md)

 - [Function constructor](./functions.md#funCtor): `[ x \in S |-> e ]`
 - [Set of functions](./functions.md#funSetCtor): `[S -> T]`
 - [Function application](./functions.md#funApp): `f[e]`
 - [Function replacement](./functions.md#except): `[ f EXCEPT ![e_1] = e_2 ]`
 - [Function domain](./functions.md#domain): `DOMAIN f`

### Records :books:

_Records like everywhere else_. [Learn more...](./records.md)

 - All operators of [functions](./functions.md)
 - [Record constructor](./records.md#recCtor): `[ h_1 |-> e_1, ..., h_n |-> e_n ]`
 - [Set of records](./records.md#recSetCtor): `[ h_1: S_1, ..., h_n: S_n ]`
 - [Access by field name](./records.md#recApp): `e.h`

### Tuples :triangular_ruler:

_Well, tuples_, indexed with 1, 2, 3... [Learn more...](./tuples.md)

>   - All operators of [functions](./functions.md)

Author: Right. It is a confusing concept. I have written literally: Records are functions, tuples are functions, sequences are functions.
